### PR TITLE
Support for stdin streams

### PR DESCRIPTION
**Pull request by pavannd1**
## Overview

- `fs.StreamingFile` interface with a `GetReader` method to return an `io.Reader`
- `virtualfs` package to implement an in-memory static version of `fs.Directory` and `fs.StreamingFile`
- `snapshot create <rootdir> --stdin-file <file name>` creates a static directory (rootdir) with a single streaming file entry (file name) and uploads it

## Issues

- [x] [683](https://github.com/kopia/kopia/issues/683)

## Test Plan

- [x] Unit test
- [x] End to End test
- [x] Manual
```
$ cat ./kopia | ./kopia snapshot create kopiadir --stdin-file=kopia
* 0 hashing, 1 hashed (32.3 MB), 0 cached (0 B), uploaded 32.3 MB, estimating...
Created snapshot with root k299834b5c887776879bf7e680eed6542 and ID 38e21994c4625dd128a463db0d6c0226 in 0s
```
```
$ ./kopia snapshot list
k10admin@mac:kopiadir
  2021-03-03 16:17:09 PST k299834b5c887776879bf7e680eed6542 32.3 MB drwxrwxrwx files:1 dirs:1 (latest-1,annual-1,monthly-1,weekly-1,daily-1,hourly-1)
```
```
$ ./kopia manifest list
7ea93a6362f673d797bf783530fd0832        165 2021-02-28 20:59:32 PST type:maintenance
38e21994c4625dd128a463db0d6c0226        603 2021-03-03 16:17:09 PST type:snapshot hostname:mac path:kopiadir username:k10admin
7a01c9fece138de6e01adc56fed96378        105 2021-03-03 16:17:09 PST type:policy hostname:mac path:kopiadir policyType:path username:k10admin
```
```
$ ./kopia policy show 7a01c9fece138de6e01adc56fed96378
Policy for k10admin@mac:kopiadir:

Retention:
  Annual snapshots:    3           (default)
  Monthly snapshots:  24           (default)
  Weekly snapshots:    4           (default)
  Daily snapshots:     7           (default)
  Hourly snapshots:   48           (default)
  Latest snapshots:   10           (default)

Files policy:
  Ignore cache directories:        true       (default)
  No ignore rules.
  Read ignore rules from files:
    .kopiaignore                   (default)
  Scan one filesystem only:       false       (default)

Error handling policy:
  Ignore file read errors:       false       (default)
  Ignore directory read errors:  false       (default)
  Ignore unknown types:           true       (default)

Scheduling policy:
  Scheduled snapshots:
    None
  Manual snapshot:            true   (defined for this target)

Compression disabled.

No actions defined.
```
```
$ ./kopia manifest show 38e21994c4625dd128a463db0d6c0226
// id: 38e21994c4625dd128a463db0d6c0226
// length: 603
// modified: 2021-03-03 16:17:09 PST
// label username:k10admin
// label hostname:mac
// label path:kopiadir
// label type:snapshot
{
  "source": {
    "host": "mac",
    "userName": "k10admin",
    "path": "kopiadir"
  },
  "description": "",
  "startTime": "2021-03-03T16:17:09.010755-08:00",
  "endTime": "2021-03-03T16:17:09.226163-08:00",
  "stats": {
    "totalSize": 32264704,
    "excludedTotalSize": 0,
    "fileCount": 1,
    "cachedFiles": 0,
    "nonCachedFiles": 1,
    "dirCount": 1,
    "excludedFileCount": 0,
    "excludedDirCount": 0,
    "ignoredErrorCount": 0,
    "errorCount": 0
  },
  "rootEntry": {
    "name": "kopiadir",
    "type": "d",
    "mode": "0777",
    "mtime": "0001-01-01T00:00:00Z",
    "obj": "k299834b5c887776879bf7e680eed6542",
    "summ": {
      "size": 32264704,
      "files": 1,
      "symlinks": 0,
      "dirs": 1,
      "maxTime": "0001-01-01T00:00:00Z",
      "numFailed": 0
    }
  }
}

### CODE REVIEW

File: fs/virtualfs/virtualfs.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+// virtualFile is an implementation of fs.StreamingFile with an io.Reader.
+type virtualFile struct {
+	virtualEntry
+	reader io.Reader
+}
+
+// GetReader returns the streaming file's reader.
+func (vf *virtualFile) GetReader(ctx context.Context) (io.Reader, error) {
+	return vf.reader, nil
```
Comment (jkowalski) [inline fs/virtualfs/virtualfs.go:98]: it should probably be an error to call this twice.

File: cli/command_snapshot_create.go [hunk 2]
```diff
@@ -168,6 +177,31 @@ func snapshotSingleSource(ctx context.Context, rep repo.RepositoryWriter, u *sna
...
+
+func snapshotStdinSource(ctx context.Context, rep repo.RepositoryWriter, u *snapshotfs.Uploader) error {
+	rootDirName := *snapshotCreateStdinDirName
+	stdinFileName := *snapshotCreateStdinFileName
+
+	log(ctx).Infof("Snapshotting stdin source as %v ...", rootDirName)
+
+	// stdin source will be snapshotted using a virtual static root directory with a single streaming file entry
+	// Create a new static directory with the given name and add a streaming file entry with os.Stdin reader
+	staticDir := virtualfs.NewStaticDirectory(rootDirName, fs.Entries{
```
Comment (jkowalski) [inline cli/command_snapshot_create.go:191]: I think we can simplify the flags a bit here:

What do you think of removing --rootdir to simplify the UX like this:

```
$ kopia snapshot create /some/dir --stdin-file=filename ?
```

Also, how about this:

```
$ kopia snapshot create /some/file --stdin
```

(this could create a single-file snapshot, not a directory-with-a-single-file snapshot, this is optional, could be in separate PR).

File: snapshot/snapshotfs/upload.go [hunk 3]
```diff
@@ -201,6 +201,43 @@ func (u *Uploader) uploadSymlinkInternal(ctx context.Context, relativePath strin
...
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get result")
+	}
+
+	de, err := newDirEntry(f, r)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create dir entry")
+	}
+
+	de.FileSize = written
```
Comment (jkowalski) [inline snapshot/snapshotfs/upload.go:233]: can we also set mtime to `clock.Now()` ?

File: tests/end_to_end_test/snapshot_create_test.go [hunk 4]
```diff
@@ -471,6 +472,79 @@ func TestSnapshotCreateAllWithManualSnapshot(t *testing.T) {
...
+	}
+
+	if _, err = w.Write(content); err != nil {
+		t.Fatalf("error writing to pipe file: %v", err)
+	}
+
+	w.Close()
+
+	// Use the pipe file as stdin
+	initialStdin := os.Stdin
```
Comment (jkowalski) [inline tests/end_to_end_test/snapshot_create_test.go:498]: this modifies global state and will potentially mess up other tests. 

Instead you can pass stdin Reader to subprocess only (exec.Cmd.Stdin), without having to change current process.

File: tests/testenv/cli_test_env.go [hunk 5]
```diff
@@ -287,6 +289,10 @@ func (e *CLITest) Run(t *testing.T, expectedError bool, args ...string) (stdout,
 		c.Stderr = os.Stderr
 	}
 
+	if e.PassthroughStdin {
```
Comment (jkowalski) [inline tests/testenv/cli_test_env.go:292]: instead of that, you can make PasshtroughStdin into a reader and do something like:

```
c.Stdin = e.NextCommandStdin
e.NextCommandStdin = nil
```

File: snapshot/snapshotfs/upload.go [hunk 6]
```diff
@@ -201,6 +201,43 @@ func (u *Uploader) uploadSymlinkInternal(ctx context.Context, relativePath strin
...
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get streaming file reader")
+	}
+
+	writer := u.repo.NewObjectWriter(ctx, object.WriterOptions{
+		Description: "STREAMFILE:" + f.Name(),
+	})
+	defer writer.Close() //nolint:errcheck
+
+	written, err := u.copyWithProgress(writer, reader, 0, f.Size())
```
Comment (jkowalski) [inline snapshot/snapshotfs/upload.go:218]: f.Size() is not known (likely zero). What does this do in terms of progress output?

**Review by jkowalski — CHANGES_REQUESTED** [review]

File: cli/command_snapshot_create.go (hunk 2 above)
Comment (pavannd1) [reply cli/command_snapshot_create.go]: Note that to support this, I have changed the args to be a list of `Strings()`. The snapshot create command will now fail at a later stage if the provided source file or directory does not exist

File: snapshot/snapshotfs/upload.go (hunk 6 above)
Comment (pavannd1) [reply snapshot/snapshotfs/upload.go:219]: I see something like this in logs:
```
 \ 1 hashing, 0 hashed (4.9 MB), 0 cached (0 B), uploaded 210 B, estimating...[STORAGE] PutBlob("s64734ac1f42cc0e46939dc1aca1945c6-s092fe2d9a70f48d6101",len=210)=<nil> took 536.342µs
 - 1 hashing, 0 hashed (27.7 MB), 0 cached (0 B), uploaded 27.7 MB, estimating...[STORAGE] PutBlob("pd0f636a786684f59e675c87f712d91a7-s092fe2d9a70f48d6101",len=27738411)=<nil> took 11.28855ms
 - 1 hashing, 0 hashed (52.8 MB), 0 cached (0 B), uploaded 52.9 MB, estimating...[STORAGE] PutBlob("pb16aff7cb127b03d47135d0a0689b163-s092fe2d9a70f48d6101",len=25170182)=<nil> took 12.801101ms
 | 1 hashing, 0 hashed (80.6 MB), 0 cached (0 B), uploaded 80.7 MB, estimating...[STORAGE] PutBlob("p612a62fa2d2fc809f1e6d019a30e6c01-s092fe2d9a70f48d6101",len=27750699)=<nil> took 15.916912ms
 / 1 hashing, 0 hashed (104.9 MB), 0 cached (0 B), uploaded 104.9 MB, estimating...[STORAGE] PutBlob("p13f4a34665dfbcc1eebfb0fa39924a8a-s092fe2d9a70f48d6101",len=24232198)=<nil> took 17.461747ms
 / 1 hashing, 0 hashed (130 MB), 0 cached (0 B), uploaded 130.1 MB, estimating...[STORAGE] PutBlob("p7f7afdc67fe4fcf6f7dbeeaa9296afd4-s092fe2d9a70f48d6101",len=25170182)=<nil> took 75.648642ms
 - 1 hashing, 0 hashed (154.3 MB), 0 cached (0 B), uploaded 154.4 MB, estimating...[STORAGE] PutBlob("p69ac3031aeff454355b323bb3ccd62a6-s092fe2d9a70f48d6101",len=24297771)=<nil> took 15.973844ms
 / 1 hashing, 0 hashed (179.5 MB), 0 cached (0 B), uploaded 179.5 MB, estimating...[STORAGE] PutBlob("pf46dc28ab08fa923ca28880bc1b5d409-s092fe2d9a70f48d6101",len=25170182)=<nil> took 20.798219ms
 / 1 hashing, 0 hashed (200.9 MB), 0 cached (0 B), uploaded 200.9 MB, estimating...[STORAGE] PutBlob("p1b9c119c0c69548d0b81f58fd8c4f18e-s092fe2d9a70f48d6101",len=21414150)=<nil> took 9.483296ms
 - 1 hashing, 0 hashed (200.9 MB), 0 cached (0 B), uploaded 200.9 MB, estimating...listing manifest contents
  keeping 2021-03-04 00:11:11.828543 -0800 PST reasons: [latest-1,annual-1,monthly-1,weekly-1,daily-1,hourly-1]
 \ 0 hashing, 1 hashed (201 MB), 0 cached (0 B), uploaded 201 MB, estimating...    [STORAGE] PutBlob("pb0ca7b2793316116d82236c202ebd69e-s092fe2d9a70f48d6101",len=86203)=<nil> took 443.026µs
 ```

**Comment by julio-lopez** [comment]
Thanks @pavannd1 !

**Comment by oderwat** [comment]
Thank you for this feature. I wonder how to restore a file to stdout to complement the process for the restoration. Am I missing something, or is this not supported yet?

**Comment by jkowalski** [comment]
If you have object ID, you can simply stream it to stdout using `kopia show <object-id>`

**Comment by jkowalski** [comment]
```
$ echo "Hello World" | kopia snapshot create /my-stdin --stdin-file=my-file
Created snapshot with root k4fd88040ad48119ebcc35d351354d3bb and ID 04451fffae90602998b375d698d7ce9e in 0s

$ kopia show k4fd88040ad48119ebcc35d351354d3bb/my-file
Hello World
```

**Comment by oderwat** [comment]
OK! That is cool! I just implemented it with a named pipe but that will make it easier!
